lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php

```
@@ -844,7 +844,11 @@ public function addDiscriminatorToPreparedQuery(array $preparedQuery)
          */
         if ($this->class->hasDiscriminator() && ! isset($preparedQuery[$this->class->discriminatorField])) {
             $discriminatorValues = $this->getClassDiscriminatorValues($this->class);
-            $preparedQuery[$this->class->discriminatorField] = array('$in' => $discriminatorValues);
+            if((count($discriminatorValues) == 1)) {
+                $preparedQuery[$this->class->discriminatorField] = $discriminatorValues[0];
+            } else {
+                $preparedQuery[$this->class->discriminatorField] = array('$in' => $discriminatorValues);
+            }
```

I think this PR needs additional checks to ensure that upserts never resort to using `$in` when dealing with discriminators. We should throw an exception that tells users they must use a specific class repository or query builder so that we do have a single discriminator to work with.

As-is, there's still an edge case where the user runs an upsert against a base class query builder and we miss including a discriminator value in the newly inserted document.